Extension/src/Utility/Text/taggedLiteral.ts

```diff
@@ -0,0 +1,254 @@
+                result.state = 'text';
+                continue;
+
+            case 'dollar':
+                if (char === '{') {
+                    result.state = 'substitution';
+                    result.template.push(template);
+                    template = '';
+                    continue;
+                }
+                template = `${template}$${char}`;
+                result.state = 'text';
+                continue;
+
+            case 'substitution':
+                switch (char) {
+                    case '}':
+                        result.expressions.push(expression);
+                        expression = '';
+                        result.state = 'text';
+                        continue;
+
+                    case ' ':
+                    case '\t':
+                    case '\r':
+                    case '\n':
+                        continue; // ignore whitespace
+
+                    case ':':
+                    case '.':
+                        expression += ':';
+                        continue;
+                }
+                if (expression) {
+                    if (isIdentifierPart(char.codePointAt(0)!) || char === '-' || char === '/') {
+                        expression += char;
+                        continue;
+                    }
+                    // error, fall thru
+                } else if (isIdentifierStart(char.codePointAt(0)!) || char === '-' || char === '/') {
+                    expression += char;
+                    continue;
+                }
+
+                // not a valid character for an expression
+                result.state = 'error';
+                result.message = `Unexpected character '${char}' in expression ${expression}`;
+                return result;
+        }
+    }
+
+    switch (result.state) {
+        case 'escape':
+            result.state = 'error';
+            result.message = 'Unexpected end of string (trailing backslash)';
+            return result;
+        case 'substitution':
+            result.state = 'error';
+            result.message = 'Unexpected end of string parsing expression ${ ';
+            return result;
+        case 'dollar':
+            template += '$';
+            break;
+    }
+    result.state = 'ok';
+    result.template.push(template);
+    return result;
+}
+
+function split(expression: string) {
+    return (expression.match(/(.*?):(.*)/) || ['', '', expression]).slice(1);
+}
+
+function resolveValue(expression: string, context: Record<string, any>, customResolver = (_prefix: string, _expression: string) => ''): string {
+    const [prefix, suffix] = split(expression);
+
+    function joinIfArray(value: any, separator = '\u0007') {
+        return Array.isArray(value) ? value.join(separator) : value.toString();
+    }
+
+    if (prefix) {
+        const variable = context[prefix];
+        if (variable !== undefined && variable !== null) {                  // did we get back an actual value

```

Comment:
This comment indentation appears incorrect. Can it be improved?

Reply:
reducing whitespace to 1

Reply:
added an eslint rule to eliminate extra spacing